File: greedy/boj1781.py
from queue import PriorityQueue 

class Pair : 
    def __init__(self, Deadline, CupRamen) : 
        self.Deadline = Deadline 
        self.CupRamen = CupRamen
    def __lt__(self, other) : 
        if(self.Deadline < other.Dealine) : 
            return True 
        elif (self.Deadline == other.Deadline) : 
            return self.CupRamen > other.CupRamen
        return False
    def __repr__(self) : 
        return '<Pair Dealine={self.Dealine} CupRamen={self.CupRamen}>'

if __name__ == "__main__" : 
    pq = PriorityQueue() 
    n = input() 
    for i in range(n) : 
        a, b = input()
        print(a, b)

File: greedy/boj1781.cpp
#include <iostream> 
#include <queue> 
#include <algorithm> 
using namespace std; 
typedef pair<int, int> D_C; 

bool DCsort(D_C left, D_C right) {
    if(left.first > right.first)
        return true; 
    else if(left.first == right.first)
        return left.second > right.second; 
    return false; 
}

int main(){
    int n; 
    cin >> n; 
    vector<D_C> dcs; 
    dcs.reserve(n);  
    for(int i = 0; i < n; i++){
        int deadline, cupRamen;
        cin >> deadline >> cupRamen;  
        dcs.emplace_back(deadline, cupRamen); 
    }
    sort(dcs.begin(), dcs.end(), DCsort); 
    //for(int i = 0; i < n; i++)
    //    cout << dcs[i].first << ", " << dcs[i].second << endl; 

    priority_queue<int, vector<int>, less<int>> pq; 
    int idx = 0; 
    int res = 0; 
    for(int day=n; day > 0; day--) {
        while(idx < n && dcs[idx].first >= day) {
            pq.push(dcs[idx].second); 
            idx++; 
        }
        if(!pq.empty()){
            res += pq.top(); 
            //cout << "day " << day << " , "<< pq.top() << endl; 
            pq.pop(); 
        }
    }
    cout << res << endl; 
    return 0; 
}